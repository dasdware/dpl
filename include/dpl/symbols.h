#ifndef __DPL_SYMBOLS_H
#define __DPL_SYMBOLS_H

#include <arena.h>
#include <nob.h>

#include <program.h>

typedef struct DPL_Symbol DPL_Symbol;

typedef enum
{
    SYMBOL_BOUNDARY,
    SYMBOL_TYPE,
    SYMBOL_CONSTANT,
    SYMBOL_VAR,
    SYMBOL_FUNCTION,
    SYMBOL_ARGUMENT,

    COUNT_SYMBOL_KINDS,
} DPL_Symbol_Kind;

// BOUNDARIES

typedef enum
{
    BOUNDARY_MODULE,
    BOUNDARY_SCOPE,
    BOUNDARY_FUNCTION,

    COUNT_SYMBOL_BOUNDARY_KINDS,
} DPL_Symbol_Boundary_Kind;

// TYPES

typedef enum
{
    TYPE_BASE,
    TYPE_FUNCTION,
    TYPE_OBJECT,
    TYPE_ALIAS,

    COUNT_SYMBOL_TYPE_KINDS,
} DPL_Symbol_Type_Kind;

typedef enum
{
    TYPE_BASE_NUMBER,
    TYPE_BASE_STRING,
    TYPE_BASE_BOOLEAN,
    TYPE_BASE_NONE,

    COUNT_SYMBOL_TYPE_BASE_KINDS,
} DPL_Symbol_Type_Base_Kind;

#define TYPENAME_NUMBER "Number"
#define TYPENAME_STRING "String"
#define TYPENAME_BOOLEAN "Boolean"
#define TYPENAME_NONE "None"

typedef struct
{
    Nob_String_View name;
    DPL_Symbol *type;
} DPL_Symbol_Type_ObjectField;

typedef da_array(DPL_Symbol_Type_ObjectField) DPL_Symbol_Type_ObjectQuery;

typedef struct
{
    size_t field_count;
    DPL_Symbol_Type_ObjectField *fields;
} DPL_Symbol_Type_Object;

typedef struct
{
    size_t argument_count;
    DPL_Symbol **arguments;
    DPL_Symbol *returns;
} DPL_Symbol_Type_Signature;

typedef struct
{
    DPL_Symbol_Type_Kind kind;
    union
    {
        DPL_Symbol_Type_Base_Kind base;
        DPL_Symbol_Type_Object object;
        DPL_Symbol_Type_Signature function;
        DPL_Symbol *alias;
    } as;
} DPL_Symbol_Type;

// Constants

typedef struct DPL_Symbol_Constant DPL_Symbol_Constant;

typedef struct
{
    Nob_String_View name;
    DPL_Symbol_Constant *value;
} DPL_Symbol_Constant_ObjectField;

typedef struct
{
    size_t field_count;
    DPL_Symbol_Constant_ObjectField *fields;
} DPL_Symbol_Constant_Object;

struct DPL_Symbol_Constant
{
    DPL_Symbol *type;
    union
    {
        double number;
        Nob_String_View string;
        bool boolean;

        DPL_Symbol_Constant_Object object;
    } as;
};

// Variables

typedef struct
{
    DPL_Symbol *type;
    size_t scope_index;
} DPL_Symbol_Var;

// Functions

typedef enum
{
    FUNCTION_INSTRUCTION,
    FUNCTION_EXTERNAL,
    FUNCTION_USER,

    COUNT_SYMBOL_FUNCTION_KINDS,
} DPL_Symbol_Function_Kind;

typedef struct
{
    DPL_Symbol_Type_Signature signature;
    DPL_Symbol_Function_Kind kind;
    union
    {
        DPL_Instruction_Kind instruction_function;
        size_t external_function;
        struct
        {
            void *body;
            bool used;
            size_t user_handle;
            size_t function_handle;
        } user_function;
    } as;
} DPL_Symbol_Function;

// Symbols

struct DPL_Symbol
{
    DPL_Symbol_Kind kind;
    Nob_String_View name;
    union
    {
        DPL_Symbol_Boundary_Kind boundary;
        DPL_Symbol_Constant constant;
        DPL_Symbol_Var var;
        DPL_Symbol_Var argument;
        DPL_Symbol_Function function;
        DPL_Symbol_Type type;
    } as;

    size_t boundary_count;
    int stack_index;
};

// Symbol stack

typedef struct
{
    DPL_Symbol **entries;
    size_t entries_count;
    size_t entries_capacity;

    Arena memory;
} DPL_SymbolStack;

// API

const char *dpl_symbols_kind_name(DPL_Symbol_Kind kind);

void dpl_symbols_init(DPL_SymbolStack *stack);
void dpl_symbols_free(DPL_SymbolStack *stack);
const char *dpl_symbols_last_error();

DPL_Symbol *dpl_symbols_find(DPL_SymbolStack *stack, Nob_String_View name);
DPL_Symbol *dpl_symbols_find_cstr(DPL_SymbolStack *stack, const char *name);
DPL_Symbol *dpl_symbols_find_kind(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol_Kind kind);
DPL_Symbol *dpl_symbols_find_kind_cstr(DPL_SymbolStack *stack, const char *name, DPL_Symbol_Kind kind);
DPL_Symbol *dpl_symbols_find_type_base(DPL_SymbolStack *stack, DPL_Symbol_Type_Base_Kind kind);
#define dpl_symbols_find_type_number(symbols) dpl_symbols_find_type_base((symbols), TYPE_BASE_NUMBER)
#define dpl_symbols_find_type_string(symbols) dpl_symbols_find_type_base((symbols), TYPE_BASE_STRING)
#define dpl_symbols_find_type_boolean(symbols) dpl_symbols_find_type_base((symbols), TYPE_BASE_BOOLEAN)
#define dpl_symbols_find_type_none(symbols) dpl_symbols_find_type_base((symbols), TYPE_BASE_NONE)
DPL_Symbol *dpl_symbols_find_type_object_query(DPL_SymbolStack *stack, DPL_Symbol_Type_ObjectQuery query);
DPL_Symbol *dpl_symbols_find_function(DPL_SymbolStack *stack, Nob_String_View name, size_t arguments_count, DPL_Symbol **arguments);
DPL_Symbol *dpl_symbols_find_function1(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol *arg0);
DPL_Symbol *dpl_symbols_find_function1_cstr(DPL_SymbolStack *stack, const char *name, DPL_Symbol *arg0);
DPL_Symbol *dpl_symbols_find_function2(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol *arg0, DPL_Symbol *arg1);
DPL_Symbol *dpl_symbols_find_function2_cstr(DPL_SymbolStack *stack, const char *name, DPL_Symbol *arg0, DPL_Symbol *arg1);

// Common
DPL_Symbol *dpl_symbols_push(DPL_SymbolStack *stack, DPL_Symbol_Kind kind, Nob_String_View name);
DPL_Symbol *dpl_symbols_push_cstr(DPL_SymbolStack *stack, DPL_Symbol_Kind kind, const char *name);

// Boundaries
DPL_Symbol *dpl_symbols_push_boundary_cstr(DPL_SymbolStack *stack, const char *name, DPL_Symbol_Boundary_Kind kind);
bool dpl_symbols_pop_boundary(DPL_SymbolStack *stack);

// Types
DPL_Symbol *dpl_symbols_push_type_base_cstr(DPL_SymbolStack *stack, const char *name, DPL_Symbol_Type_Base_Kind base_kind);
DPL_Symbol *dpl_symbols_push_type_object_cstr(DPL_SymbolStack *stack, const char *name, size_t field_count);
DPL_Symbol *dpl_symbols_push_type_alias(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol *type);

bool dpl_symbols_is_type_base(DPL_Symbol *symbol, DPL_Symbol_Type_Base_Kind kind);

// Constants
DPL_Symbol *dpl_symbols_push_constant_number_cstr(DPL_SymbolStack *stack, const char *name, double value);

// Functions
DPL_Symbol *dpl_symbols_push_function_instruction_cstr(DPL_SymbolStack *stack,
                                                       const char *name, const char *return_type, size_t argument_count, const char **argument_types,
                                                       DPL_Instruction_Kind instruction);
DPL_Symbol *dpl_symbols_push_function_external_cstr(DPL_SymbolStack *stack,
                                                    const char *name, const char *return_type, size_t argument_count, const char **argument_types,
                                                    size_t index);
DPL_Symbol *dpl_symbols_push_function_user(DPL_SymbolStack *stack, Nob_String_View name, size_t argument_count);

// Variables
DPL_Symbol *dpl_symbols_push_var(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol *type);
DPL_Symbol *dpl_symbols_push_var_cstr(DPL_SymbolStack *stack, const char *name, const char *type);

// Arguments
DPL_Symbol *dpl_symbols_push_argument(DPL_SymbolStack *stack, Nob_String_View name, DPL_Symbol *type);
DPL_Symbol *dpl_symbols_push_argument_cstr(DPL_SymbolStack *stack, const char *name, const char *type_name);

// Printing
void dpl_symbols_print_signature(DPL_Symbol *symbol, Nob_String_Builder *sb);
void dpl_symbols_print_full_signature(DPL_Symbol *symbol, Nob_String_Builder *sb);
void dpl_symbols_print_table(DPL_SymbolStack *stack);

#endif // __DPL_SYMBOLS_H