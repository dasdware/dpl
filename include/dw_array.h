#ifndef DW_ARRAY_H_INCLUDED
#define DW_ARRAY_H_INCLUDED

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#ifndef DA_INITIAL_CAPACITY
#   define DA_INITIAL_CAPACITY 8
#endif

#ifndef DA_MALLOC
#  define DA_REALLOC realloc
#endif

#ifndef DA_FREE
#  define DA_FREE free
#endif

#define da_array(type) type*

typedef struct {
    size_t size;
    size_t capacity;
    char data[];
} DA_Header;

size_t da_size(void* array);
#define da_empty(array) (da_size(array) == 0)
#define da_some(array) (da_size(array) > 0)

void da_set_size(void *array, size_t new_size);
void da_pop(void *array);
#define da_clear(array) da_set_size(array,  0)

#define da_check_size(array, size)                           \
    do {                                                     \
        _da_check_capacity((array), sizeof(*(array)), size); \
        da_set_size((array), size);                          \
    } while(false)

size_t da_capacity(void *array);

void da_free(void* array);

#define _da_check_capacity(array, element_size, n)                                                           \
    do {                                                                                                     \
        bool old_exists = ((array) != NULL);                                                                 \
        size_t old_capacity = da_capacity(array);                                                            \
                                                                                                             \
        size_t new_size = da_size(array) + n;                                                                \
        size_t new_capacity = (old_exists) ? old_capacity : DA_INITIAL_CAPACITY;                             \
        while (new_size > new_capacity) {                                                                    \
            new_capacity *= 2;                                                                               \
        }                                                                                                    \
                                                                                                             \
        if (new_capacity > old_capacity) {                                                                   \
            DA_Header* old_header = (old_exists) ? ((DA_Header*) (array)) - 1 : NULL;                        \
            DA_Header* new_header = DA_REALLOC(old_header, sizeof(DA_Header) + new_capacity * element_size); \
            if (!old_exists) {                                                                               \
                new_header->size = 0;                                                                        \
            }                                                                                                \
            new_header->capacity = new_capacity;                                                             \
            (array) = (void*) &(new_header[1]);                                                              \
        }                                                                                                    \
    } while (false)


#define da_add(array, element)                                    \
    do {                                                          \
        size_t array_size = da_size((array));                     \
        _da_check_capacity((array), sizeof(*(array)), 1);         \
        (array)[array_size] = element;                            \
        ((DA_Header*) array)[-1].size++;                          \
    } while(false)

#define da_addn(array, elements, count)                                     \
    do {                                                                    \
        size_t array_size = da_size((array));                               \
        _da_check_capacity((array), sizeof(*(array)), count);               \
        memcpy((array) + array_size, (elements), count * sizeof(*(array))); \
        ((DA_Header*) array)[-1].size += count;                             \
    } while(false)


typedef da_array(char) str_t;

#define str_length(str) da_size(str)

#define str_append(string, text)                                    \
    do {                                                            \
        size_t len = str_length((string));                          \
        size_t count = strlen(text);                                \
        _da_check_capacity((string), sizeof(*(string)), count + 1); \
        memcpy((string) + len, (text), count * sizeof(*(string)));  \
        (string)[len + count] = '\0';                               \
        ((DA_Header*) string)[-1].size += count;                    \
    } while(false)

#define str_append_length(string, data, count)                      \
    do {                                                            \
        size_t len = str_length((string));                          \
        _da_check_capacity((string), sizeof(*(string)), count + 1); \
        memcpy((string) + len, (data), count * sizeof(*(string)));  \
        (string)[len + count] = '\0';                               \
        ((DA_Header*) string)[-1].size += count;                    \
    } while(false)


str_t str_new(const char* text);
#define str_free(string) da_free(string)

str_t str_empty();

#endif // DW_ARRAY_H_INCLUDED

#ifdef DW_ARRAY_IMPLEMENTATION
#undef DW_ARRAY_IMPLEMENTATION

size_t da_size(void* array) {
    if (!array) {
        return 0;
    }
    return ((DA_Header*) array)[-1].size;
}

void da_set_size(void *array, size_t new_size) {
    if (new_size < da_capacity(array))  {
        ((DA_Header*) array)[-1].size = new_size;
    }
}

void da_pop(void *array) {
    if (da_size(array) > 0)  {
        ((DA_Header*) array)[-1].size--;
    }
}


size_t da_capacity(void *array) {
    if (!array) {
        return 0;
    }
    return ((DA_Header*) array)[-1].capacity;
}

void da_free(void* array) {
    if (array) {
        DA_FREE(((DA_Header*) array) - 1);
    }
}

str_t str_new(const char* text) {
    str_t result = 0;
    str_append(result, text);
    return result;
}

str_t str_empty() {
    return str_new("");
}

#endif // DW_ARRAY_IMPLEMENTATION