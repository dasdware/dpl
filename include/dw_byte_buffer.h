#ifndef DW_BYTE_BUFFER_H_INCLUDED
#define DW_BYTE_BUFFER_H_INCLUDED

typedef struct {
    uint8_t *items;
    size_t count;
    size_t capacity;
} DW_ByteBuffer;

void bb_write_u8(DW_ByteBuffer *buffer, uint8_t value);
void bb_write_u16(DW_ByteBuffer *buffer, uint16_t value);
void bb_write_u32(DW_ByteBuffer *buffer, uint32_t value);
void bb_write_u64(DW_ByteBuffer *buffer, uint64_t value);
void bb_write_f64(DW_ByteBuffer *buffer, double value);
void bb_write_sv(DW_ByteBuffer *buffer, Nob_String_View value);

uint8_t bb_read_u8(DW_ByteBuffer buffer, size_t offset);
uint16_t bb_read_u16(DW_ByteBuffer buffer, size_t offset);
uint32_t bb_read_u32(DW_ByteBuffer buffer, size_t offset);
uint64_t bb_read_u64(DW_ByteBuffer buffer, size_t offset);
double bb_read_f64(DW_ByteBuffer buffer, size_t offset);
Nob_String_View bb_read_sv(DW_ByteBuffer buffer, size_t offset);

void bb_save(FILE *out, DW_ByteBuffer buffer);

typedef struct _DW_ByteStream
{
    DW_ByteBuffer buffer;
    size_t position;
} DW_ByteStream;

bool bs_at_end(DW_ByteStream* stream);

uint8_t bs_read_u8(DW_ByteStream* stream);
uint16_t bs_read_u16(DW_ByteStream* stream);
uint32_t bs_read_u32(DW_ByteStream* stream);
uint64_t bs_read_u64(DW_ByteStream* stream);
double bs_read_f64(DW_ByteStream* stream);

#endif // DW_ARRAY_H_INCLUDED

#ifdef DW_BYTEBUFFER_IMPLEMENTATION
#undef DW_BYTEBUFFER_IMPLEMENTATION

void bb_write_u8(DW_ByteBuffer *buffer, uint8_t value)
{
    nob_da_append_many(buffer, &value, sizeof(value));
}

void bb_write_u16(DW_ByteBuffer *buffer, uint16_t value)
{
    nob_da_append_many(buffer, &value, sizeof(value));
}

void bb_write_u32(DW_ByteBuffer *buffer, uint32_t value)
{
    nob_da_append_many(buffer, &value, sizeof(value));
}

void bb_write_u64(DW_ByteBuffer *buffer, uint64_t value)
{
    nob_da_append_many(buffer, &value, sizeof(value));
}

void bb_write_f64(DW_ByteBuffer *buffer, double value)
{
    nob_da_append_many(buffer, &value, sizeof(value));
}

void bb_write_sv(DW_ByteBuffer *buffer, Nob_String_View value)
{
    bb_write_u64(buffer, value.count);
    nob_da_append_many(buffer, value.data, sizeof(char) * value.count);
}

uint8_t bb_read_u8(DW_ByteBuffer buffer, size_t offset)
{
    return *(uint8_t*) &buffer.items[offset];
}

uint16_t bb_read_u16(DW_ByteBuffer buffer, size_t offset)
{
    return *(uint32_t*) &buffer.items[offset];
}

uint32_t bb_read_u32(DW_ByteBuffer buffer, size_t offset)
{
    return *(uint32_t*) &buffer.items[offset];
}

uint64_t bb_read_u64(DW_ByteBuffer buffer, size_t offset)
{
    return *(uint64_t*) &buffer.items[offset];
}

double bb_read_f64(DW_ByteBuffer buffer, size_t offset)
{
    return *(double*) &buffer.items[offset];
}

Nob_String_View bb_read_sv(DW_ByteBuffer buffer, size_t offset)
{
    return (Nob_String_View) {
        .count = bb_read_u64(buffer, offset),
        .data = (const char*) &buffer.items[offset + sizeof(uint64_t)],
    };
}

void bb_save(FILE *out, DW_ByteBuffer buffer)
{
    size_t buffer_size = buffer.count;
    fwrite(&buffer_size, sizeof(size_t), 1, out);
    fwrite(buffer.items, sizeof(uint8_t), buffer_size, out);
}

bool bs_at_end(DW_ByteStream* stream)
{
    return stream->position >= stream->buffer.count;
}

uint8_t bs_read_u8(DW_ByteStream* stream) {
    size_t offset = stream->position;
    stream->position += sizeof(uint8_t);
    return bb_read_u8(stream->buffer, offset);
}

uint16_t bs_read_u16(DW_ByteStream* stream) {
    size_t offset = stream->position;
    stream->position += sizeof(uint16_t);
    return bb_read_u16(stream->buffer, offset);
}

uint32_t bs_read_u32(DW_ByteStream* stream) {
    size_t offset = stream->position;
    stream->position += sizeof(uint32_t);
    return bb_read_u32(stream->buffer, offset);
}

uint64_t bs_read_u64(DW_ByteStream* stream) {
    size_t offset = stream->position;
    stream->position += sizeof(uint64_t);
    return bb_read_u64(stream->buffer, offset);
}

double bs_read_f64(DW_ByteStream* stream) {
    size_t offset = stream->position;
    stream->position += sizeof(double);
    return bb_read_f64(stream->buffer, offset);
}

#endif // DW_BYTE_BUFFER_H_INCLUDED