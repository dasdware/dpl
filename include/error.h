#ifndef __DW_ERROR_H
#define __DW_ERROR_H

#include <stdio.h>

#ifndef DW_ERROR_STREAM
#define DW_ERROR_STREAM stderr
#endif

#ifndef DW_ERROR_EXIT_CODE
#define DW_ERROR_EXIT_CODE 1
#endif

#define DW_ERROR_MSG(format, ...)                        \
    do                                                   \
    {                                                    \
        fprintf(DW_ERROR_STREAM, format, ##__VA_ARGS__); \
    } while (false)

#define DW_ERROR_MSGLN(format, ...)                      \
    do                                                   \
    {                                                    \
        fprintf(DW_ERROR_STREAM, format, ##__VA_ARGS__); \
        fprintf(DW_ERROR_STREAM, "\n");                  \
    } while (false)

#define DW_ERROR(format, ...)                            \
    do                                                   \
    {                                                    \
        DW_ERROR_MSGLN(format, ##__VA_ARGS__);           \
        exit(DW_ERROR_EXIT_CODE);                        \
    } while (false)

#define DW_UNIMPLEMENTED                                 \
    DW_ERROR("\n\n%s:%d: UNIMPLEMENTED (function %s)", __FILE__, __LINE__, __FUNCTION__)

#define DW_UNIMPLEMENTED_MSG(format, ...)               \
    DW_ERROR("\n\n%s:%d: UNIMPLEMENTED (function %s): " format, __FILE__, __LINE__, __FUNCTION__, ##__VA_ARGS__)

#define DW_UNUSED(x) ((void) x)


#endif // __DW_ERROR_H