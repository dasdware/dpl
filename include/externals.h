#ifndef __DPL_EXTERNALS_H
#define __DPL_EXTERNALS_H

#include "nob.h"
#include "vm.h"

typedef const char *DPL_External_TypeName;

typedef struct
{
    DPL_External_TypeName *items;
    size_t count;
    size_t capacity;
} DPL_External_TypeNames;

typedef void (*DPL_ExternalFunction_Callback)(DPL_VirtualMachine *vm);

typedef struct
{
    const char *name;
    DPL_External_TypeNames argument_types;
    DPL_External_TypeName return_type;
    DPL_ExternalFunction_Callback callback;

} DPL_ExternalFunction;

typedef struct
{
    DPL_ExternalFunction *items;
    size_t count;
    size_t capacity;
} DPL_ExternalFunctions;

void dple_init(DPL_ExternalFunctions *externals);
void dple_free(DPL_ExternalFunctions *externals);

DPL_ExternalFunction *dple_add_by_name(DPL_ExternalFunctions *externals, const char *name);

#endif // __DPL_EXTERNALS_H