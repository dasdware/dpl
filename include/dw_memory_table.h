#ifndef DW_MEMORY_TABLE_H_INCLUDED
#define DW_MEMORY_TABLE_H_INCLUDED

// This is an stb-style library, use the following to include method definitions:
//
// #define DW_MEMORY_TABLE_IMPLEMENTATION
// #include <dw_memory_table.h>
//
// The header <ring_buffer.h> must be on the include path for definitions to work.

#include <nob.h>

#ifndef MT_CAPACITY
#define MT_CAPACITY 32ull
#endif

#ifndef MT_MAX_LENGTH
#define MT_MAX_LENGTH 256ull
#endif

typedef struct
{
    bool is_free;
    size_t length;
    char data[sizeof(size_t) + MT_MAX_LENGTH + 1];
} DW_MemoryTable_Item;

typedef struct
{
    size_t *items;

    size_t capacity;
    size_t start;
    size_t count;
} DW_MemoryTable_FreeItems;

typedef struct
{
    DW_MemoryTable_Item *items;
    DW_MemoryTable_FreeItems free_items;
} DW_MemoryTable;

void mt_init(DW_MemoryTable *table);
void mt_free(DW_MemoryTable *table);

Nob_String_View mt_sv_allocate(DW_MemoryTable *table, size_t length);
Nob_String_View mt_sv_allocate_cstr(DW_MemoryTable *table, const char *value);
Nob_String_View mt_sv_allocate_lstr(DW_MemoryTable *table, const char *data, size_t length);
Nob_String_View mt_sv_allocate_sv(DW_MemoryTable *table, Nob_String_View sv);
Nob_String_View mt_sv_allocate_concat(DW_MemoryTable *table, Nob_String_View handle1, Nob_String_View handle2);

void mt_sv_release(DW_MemoryTable *table, Nob_String_View handle);

#endif // DW_MEMORY_TABLE_H_INCLUDED

#ifdef DW_MEMORY_TABLE_IMPLEMENTATION

#include <stdio.h>
#include <string.h>

#include <error.h>
#include <ring_buffer.h>

void mt_init(DW_MemoryTable *table)
{
    table->items = calloc(MT_CAPACITY, sizeof(DW_MemoryTable_Item));

    table->free_items.capacity = MT_CAPACITY;
    rb_init(table->free_items);

    for (size_t handle = 0; handle < MT_CAPACITY; ++handle)
    {
        table->items[handle].is_free = true;
        *((size_t*)(table->items[handle].data)) = handle;

        rb_enqueue(table->free_items, handle);
    }
}

void mt_free(DW_MemoryTable *table)
{
    rb_free(table->free_items);
    free(table->items);
}

Nob_String_View mt_sv_allocate(DW_MemoryTable *table, size_t length)
{
    if (length >= MT_MAX_LENGTH)
    {
        DW_ERROR("ERROR: Cannot allocate new string from table: Requested size (%zu) is longer than accepted maximum (%llu).", length, MT_MAX_LENGTH);
    }
    if (table->free_items.count == 0)
    {
        DW_ERROR("ERROR: Cannot allocate new string from table: Table is full.");
    }

    size_t handle;
    rb_dequeue(table->free_items, handle);

    table->items[handle].is_free = false;
    table->items[handle].length = length;
    table->items[handle].data[sizeof(size_t) + length] = '\0';

    return nob_sv_from_parts(table->items[handle].data + sizeof(size_t), length);
}

Nob_String_View mt_sv_allocate_cstr(DW_MemoryTable *table, const char *value)
{
    size_t length = strlen(value);

    Nob_String_View result = mt_sv_allocate(table, length);
    memcpy((char*)result.data, value, length);

    return result;
}

Nob_String_View mt_sv_allocate_lstr(DW_MemoryTable *table, const char *data, size_t length)
{
    Nob_String_View result = mt_sv_allocate(table, length);
    memcpy((char*)result.data, data, length);

    return result;
}

Nob_String_View mt_sv_allocate_sv(DW_MemoryTable *table, Nob_String_View sv) {
    return mt_sv_allocate_lstr(table, sv.data, sv.count);
}

Nob_String_View mt_sv_allocate_concat(DW_MemoryTable *table, Nob_String_View sv1, Nob_String_View sv2)
{
    Nob_String_View result = mt_sv_allocate(table, sv1.count + sv2.count);
    memcpy((char*)result.data, sv1.data, sv1.count);
    memcpy((char*)result.data + sv1.count, sv2.data, sv2.count);

    return result;
}


size_t _mt_check_handle_value(DW_MemoryTable *table, Nob_String_View sv, const char* operation)
{
    size_t handle = *((size_t*)((sv.data) - sizeof(size_t)));

    if (handle >= MT_CAPACITY)
    {
        DW_ERROR("ERROR: Cannot %s: Handle out of range (%zu >= %llu).", operation, handle, MT_CAPACITY);
    }
    if (table->items[handle].is_free)
    {
        DW_ERROR("ERROR: Cannot %s: Entry already is released (%zu).", operation, handle);
    }

    return handle;
}

void mt_sv_release(DW_MemoryTable *table, Nob_String_View sv)
{
    size_t handle = _mt_check_handle_value(table, sv, "release string");

    rb_enqueue(table->free_items, handle);
    table->items[handle].is_free = true;
}


#endif
