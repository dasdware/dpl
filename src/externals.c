#include "error.h"
#include "externals.h"

DPL_ExternalFunction* dple_add_by_name(DPL_ExternalFunctions* externals, const char* name)
{
    DPL_ExternalFunction external = {
        .name = name,
        .argument_types = {0},
        .return_type = NULL,
        .callback = NULL,
    };

    size_t index = externals->count;
    nob_da_append(externals, external);
    return &externals->items[index];
}

void _dple_print_callback(DPL_VirtualMachine* vm)
{
    DPL_Value value = dplv_peek(vm);
    switch (value.kind) {
    case VALUE_NUMBER:
        printf("%s", dpl_value_format_number(value));
        break;
    case VALUE_STRING:
        printf(SV_Fmt, SV_Arg(value.as.string));
        break;
    default:
        DW_ERROR("ERROR: `print` function callback cannot print values of kind `%s`.", dpl_value_kind_name(value.kind));
    }
}

void _dple_length_string_callback(DPL_VirtualMachine* vm) {
    DPL_Value value = dplv_peek(vm);
    dplv_return_number(vm, 1, value.as.string.count);
}

void _dple_to_string_number_callback(DPL_VirtualMachine* vm) {
    DPL_Value value = dplv_peek(vm);
    dplv_return_string(vm, 1,
                       st_allocate_cstr(&vm->strings, dpl_value_format_number(value)));
}


void dple_init(DPL_ExternalFunctions* externals)
{
    DPL_ExternalFunction* print_number = dple_add_by_name(externals, "print");
    nob_da_append(&print_number->argument_types, "number");
    print_number->return_type = "number";
    print_number->callback = _dple_print_callback;

    DPL_ExternalFunction* print_string = dple_add_by_name(externals, "print");
    nob_da_append(&print_string->argument_types, "string");
    print_string->return_type = "string";
    print_string->callback = _dple_print_callback;

    DPL_ExternalFunction* length_string = dple_add_by_name(externals, "length");
    nob_da_append(&length_string->argument_types, "string");
    length_string->return_type = "number";
    length_string->callback = _dple_length_string_callback;

    DPL_ExternalFunction* to_string_number = dple_add_by_name(externals, "to_string");
    nob_da_append(&to_string_number->argument_types, "number");
    to_string_number->return_type = "string";
    to_string_number->callback = _dple_to_string_number_callback;
}

void dple_free(DPL_ExternalFunctions *externals)
{
    nob_da_free(*externals);
}