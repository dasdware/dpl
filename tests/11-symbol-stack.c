#include <stdio.h>
#include <dpl/symbols.h>

#define ARENA_IMPLEMENTATION
#include "arena.h"

#define NOB_IMPLEMENTATION
#include "nob.h"
#undef NOB_IMPLEMENTATION

#define DW_MEMORY_TABLE_IMPLEMENTATION
#include <dw_memory_table.h>

#define DW_BYTEBUFFER_IMPLEMENTATION
#include <dw_byte_buffer.h>

#define DW_ARRAY_IMPLEMENTATION
#include <dw_array.h>

#include "externals.h"

#include <error.h>

void test_find_symbol(DPL_SymbolStack *stack, const char *name)
{
    DPL_Symbol *symbol = dpl_symbols_find_cstr(stack, name);
    if (!symbol)
    {
        printf("%s\n", dpl_symbols_last_error());
    }
    else
    {
        Nob_String_Builder sb = {0};
        dpl_symbols_print_full_signature(symbol, &sb);
        nob_sb_append_null(&sb);

        printf("Found %s successfully.\n", sb.items);

        nob_sb_free(sb);
    }
}

int main()
{
    DPL_SymbolStack stack = {0};
    dpl_symbols_init(&stack);

    dpl_symbols_push_boundary_cstr(&stack, NULL, BOUNDARY_SCOPE);

    dpl_symbols_push_constant_number_cstr(&stack, "PI", 3.14159);
    dpl_symbols_push_var_cstr(&stack, "x", "String");

    DPL_ExternalFunctions externals = NULL;
    dple_init(&externals);

    for (size_t i = 0; i < da_size(externals); ++i)
    {
        DPL_Symbol *function = dpl_symbols_push_function_external_cstr(
            &stack,
            externals[i].name, externals[i].return_type,
            da_size(externals[i].argument_types), externals[i].argument_types,
            i);

        if (!function)
        {
            Nob_String_Builder sb = {0};
            nob_sb_append_cstr(&sb, "Cannot add external function `");
            nob_sb_append_cstr(&sb, externals[i].name);
            nob_sb_append_cstr(&sb, "(");

            for (size_t j = 0; j < da_size(externals[i].argument_types); ++j)
            {
                if (j > 0)
                {
                    nob_sb_append_cstr(&sb, ", ");
                }
                nob_sb_append_cstr(&sb, externals[i].argument_types[j]);
            }
            nob_sb_append_cstr(&sb, "): ");
            nob_sb_append_cstr(&sb, externals[i].return_type);
            nob_sb_append_cstr(&sb, "`: ");
            nob_sb_append_cstr(&sb, dpl_symbols_last_error());

            nob_sb_append_null(&sb);

            DW_ERROR("%s", sb.items);
        }
    }

    dpl_symbols_push_boundary_cstr(&stack, NULL, BOUNDARY_FUNCTION);
    dpl_symbols_push_argument_cstr(&stack, "foo", "Boolean");
    dpl_symbols_push_argument_cstr(&stack, "bar", "Boolean");
    dpl_symbols_push_var_cstr(&stack, "baz", "Boolean");

    test_find_symbol(&stack, "PI");
    test_find_symbol(&stack, "bar");

    test_find_symbol(&stack, "x");
    dpl_symbols_pop_boundary(&stack);
    test_find_symbol(&stack, "x");

    dpl_symbols_print_table(&stack);

    dpl_symbols_free(&stack);
}