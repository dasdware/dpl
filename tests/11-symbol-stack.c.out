Found constant PI: Number successfully.
Found argument bar: Boolean successfully.
Cannot find symbol  `x`.
Found variable x: String successfully.
|   1 |  -1 | boundary |                                          | module                         | 
|   2 |  -1 | type     | Number                                   | base: Number                   | 
|   3 |  -1 | type     | String                                   | base: String                   | 
|   4 |  -1 | type     | Boolean                                  | base: Boolean                  | 
|   1 |  -1 | boundary |                                          | scope                          | 
|   2 |  -1 | constant | PI: Number                               | value: 3.141590                | 
|   3 |   0 | variable | x: String                                | scope_index: 0                 | 
